cmd/exit_broadcast.go

```diff
@@ -75,66 +108,137 @@ func runBcastFullExit(ctx context.Context, config exitConfig) error {
+		return "", errors.Wrap(err, "cannot decode public key hex from file name")
 	}
-
-	pubkey, err := tblsconv.PubkeyFromBytes(rawPkBytes)
+	validatorPubKey, err := core.PubKeyFromBytes(validatorPubKeyBytes)
 	if err != nil {
-		return errors.Wrap(err, "could not convert validator key bytes to BLS public key")
+		return "", errors.Wrap(err, "cannot decode core public key from hex")
 	}
 
-	// parse signature
-	signature, err := tblsconv.SignatureFromBytes(fullExit.Signature[:])
-	if err != nil {
-		return errors.Wrap(err, "could not parse BLS signature from bytes")
-	}
+	return validatorPubKey, nil
+}
 
-	exitRoot, err := sigDataForExit(
-		ctx,
-		*fullExit.Message,
-		eth2Cl,
-		fullExit.Message.Epoch,
-	)
-	if err != nil {
-		return errors.Wrap(err, "cannot calculate hash tree root for exit message for verification")
-	}
+func fetchFullExit(ctx context.Context, exitFilePath string, config exitConfig, cl *manifestpb.Cluster, identityKey *k1.PrivateKey, validatorPubKey string) (eth2p0.SignedVoluntaryExit, error) {
+	var fullExit eth2p0.SignedVoluntaryExit
+	var err error
 
-	if err := tbls.Verify(pubkey, exitRoot[:], signature); err != nil {
-		return errors.Wrap(err, "exit message signature not verified")
+	if len(exitFilePath) != 0 {
+		log.Info(ctx, "Retrieving full exit message from path", z.Str("path", exitFilePath))
+		fullExit, err = exitFromPath(exitFilePath)
+	} else {
+		log.Info(ctx, "Retrieving full exit message from publish address")
+		fullExit, err = exitFromObolAPI(ctx, validatorPubKey, config.PublishAddress, config.PublishTimeout, cl, identityKey)
 	}
 
-	if err := eth2Cl.SubmitVoluntaryExit(ctx, &fullExit); err != nil {
-		return errors.Wrap(err, "could not submit voluntary exit")
+	return fullExit, err
+}
+
+func broadcastExitsToBeacon(ctx context.Context, eth2Cl eth2wrap.Client, exits map[core.PubKey]eth2p0.SignedVoluntaryExit) error {
+	for validator, fullExit := range exits {
+		valCtx := log.WithCtx(ctx, z.Str("validator", validator.String()))
+
+		rawPkBytes, err := validator.Bytes()
+		if err != nil {
+			return errors.Wrap(err, "could not serialize validator key bytes")
+		}
+
+		pubkey, err := tblsconv.PubkeyFromBytes(rawPkBytes)
+		if err != nil {
+			return errors.Wrap(err, "could not convert validator key bytes to BLS public key")
+		}
+
+		// parse signature
+		signature, err := tblsconv.SignatureFromBytes(fullExit.Signature[:])
+		if err != nil {
+			return errors.Wrap(err, "could not parse BLS signature from bytes")
+		}
+
+		exitRoot, err := sigDataForExit(
+			valCtx,
+			*fullExit.Message,
+			eth2Cl,
+			fullExit.Message.Epoch,
+		)
+		if err != nil {
+			return errors.Wrap(err, "cannot calculate hash tree root for exit message for verification")
+		}
+
+		if err := tbls.Verify(pubkey, exitRoot[:], signature); err != nil {
+			return errors.Wrap(err, "exit message signature not verified")
+		}
+
+		if err := eth2Cl.SubmitVoluntaryExit(valCtx, &fullExit); err != nil {
```

Comment:
We may exhaust the BN here on some big clusters. I'm thinking of ways to prevent that and I'm open to suggestions.

Reply:
I think a simple delay (maybe 100ms) should be safe enough? Otherwise NOP shall tweak rate limiting on its BN.

Reply:
@gsora any view on that? I think of maybe adding a flag of type "be-gentle-with-bn" that adds some delay between broadcasts :)